### ISSUE TITLE
Add tags to Spans

### ISSUE THREAD

**Issue by felixbarny** [elastic/apm-server#1152]
Just like Transactions, Spans should have tags. This is required to be OpenTracing compatible.

Related: https://github.com/elastic/apm-server/issues/828

**Comment by simitt** [comment]
Allowing `dynamically mapped fields` for `tags` leads to every `tag` being indexed as a separate `field`. I quickly discussed some potential implications with @danielmitterdorfer:

Customer set `tags` could potentially lead to _sparse_ fields, as they might not always be set, or have different keys. From ES 6.0 on sparse fields are way more efficient (see [blog post](https://www.elastic.co/blog/elasticsearch-6-0-0-released)).

The second issue with customer set `tags` is that they could lead to _field explosion_. ES offers some [settings](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/mapping.html#mapping-limit-settings) to address this. The default value for `index.mapping.total_fields.limit` is `1000`. Having lots of fields indexed leads to higher memory overhead and to potential memory issues. Even if the limit is not hit, a large amount of fields could have bad implications on memory usage. 
APM stores all _transactions, spans_ and _errors_  into their separate indices, but for all instrumented _services_ and not split by _service_. Thus, if lots of different `tags` are used in different services, this could lead to a large amount of fields per index, resulting in hitting the fields limit. ([Interesting discussion about hitting the limit](https://discuss.elastic.co/t/limit-of-total-fields-1000-in-index-has-been-exceeded/87280/7).)

If we want to enable `span.tags` the agent-devs would need to ensure proper documentation, e.g. explaining that setting a tag per _user_id_, per query_string param or something similar would probably lead to failures for the whole Elastic APM setup.

As `span.tags` is not the first key-value construct where we allow dynamic indexing, we are already open to that issue with `transaction.tags` and `transaction.marks`, and from 6.4 on with `metrics`. 

Another option would be to create indices per services. We haven't discussed potential advantages and implications of this yet, but might also have other drawbacks.

**Comment by felixbarny** [comment]
I suggest adding a warning to the `Span#addTag` method on the agent API. The Java agent currently has this warning for the  `Transaction#addTag` API method:

```
Note: the tags are indexed in Elasticsearch so that they are searchable and aggregatable.
By all means,
you should avoid that user-specified data,
like URL parameters,
is used as a tag key as it can lead to mapping explosions.
```

**Comment by simitt** [comment]
It worries me that one misconfiguration of a developer could lead to not having stored any APM data anymore. I simulated the behaviour when the limit is reached. The APM Server still returns a `202` in this case, as the pushing to ES happens async. The APM Server only logs a warning in this case, see according part in [libbeat](https://github.com/elastic/beats/blob/master/libbeat/outputs/elasticsearch/client.go#L520). This again leads to the unpublished events NOT filling up the internal memory queue and the server keeps responding with 202s and everything looks fine. 

E.g. one developer in a company doesn't read the agent docs, but just copy & pastes some setup for adding tags from another codebase, he might not be aware of the implications. Elastic APM stops working, without any failing requests from the agents to the server. Only the server logs show what is going on. Every day a new index would be created with data until the amount of fields is exceeded. 

I strongly recommend to take more time to investigate and potentially address those issues and questions before adding `spans.tags`.

**Comment by felixbarny** [comment]
The APM server could send metrics about successful and failed indexing attempts to a metric index. That way users could at least see that something is going wrong and then have a look at the server logs. We could also pre-configure an ML job or a simple watch (rejections>successes) and notify users about such errors.

Another idea would be to let the UI detect mapping explosions (field count close to the limit) and show an appropriate error.

However, I would not reject a feature, because there is a slight chance that users could shoot themselves in the foot with it. Especially, as we already offer that feature for transactions. I don't see that the chances of a field explosion are considerably higher when we allow tags for spans.

But I'm definitely with you that we should at least make these situations easier to debug.

A more restrictive alternative would be to make it configurable in the APM server (or in the agents) which tags are allowed.

**Comment by roncohen** [comment]
> I don't see that the chances of a field explosion are considerably higher when we allow tags for spans.

Generally I agree with this, but I'm also fine to push it to 6.5 @simitt if you feel like we could find a way to avoid the foot gun.

**Comment by simitt** [comment]
Those are all valid suggestions, and they all mean some effort. That is why I suggest to take some more time to investigate and come up with a proper solution before adding the `spans.tags`.

**Comment by simitt** [comment]
@roncohen I don't have any good idea on how to avoid potential field explosion, and I don't know if we can come up with something better for the next release.
I believe that we could come up with some better support around it, like how do we handle those failures in the Server and how to make it more visible to customers that it could lead to an issue. It might also be worth to investigate if/how one could set up indices per service if necessary and add a section in the docs on how to overcome issues. 

The implementation to support tags on spans is ready and could be merged. I feel this is a product decision that needs to be made, if we want to add the feature now and take the risk or not.

**Comment by roncohen** [comment]
OK. In that case I think we should move forward with this for 6.4  because:

* we already support tags in transactions, so in a sense we're not taking on additional risk by allowing tags in spans
* index per service will not help in what i think is the most likely scenario (timestamp tag name or similar)
* if we added a whitelist for tag names on the APM Server it's not clear to me at the moment if anyone would use it. If it turns out to be a solution that people want, we can add a whitelist for tag names on APM Server at a later point.
* I don't see a way to propagate the response from ES back to agents with our current architecture

I definitely agree that we need good documentation to support this, so if we don't have time to document it well, we should wait to 6.5.

**Comment by alvarolobato** [comment]
I agree with @roncohen lets do it for this version and think on a better solution longterm.
There are other options that we could investigate like nested type / parent-child relation although not fully supported out of the box by Kibana yet,  but I think they are working on it and in any case we would have to do it for 7.0.

The proper documentation is a must, @felixbarny do you think you could have a look at what we currently have and propose the changes?, I think that something in the line of what you already had should be enough.

**Comment by felixbarny** [comment]
I think we should add a section to the troubleshooting guide on how to debug this issue. Just by describing it and mentioning the error messages should make it easier to understand the root cause for users googling for the error message.

@simitt do you still have the errors at hand?

**Comment by simitt** [comment]
```
{"level":"warn","timestamp":"2018-07-19T11:59:21.228Z","caller":"elasticsearch/client.go:520","message":"Cannot index event publisher.Event{...}, Flags:0x1} (status=400): {\"type\":\"illegal_argument_exception\",\"reason\":\"Limit of total fields [20] in index [apm-7.0.0-alpha1-transaction-2017.05.30] has been exceeded\"}
"}
```

**Comment by simitt** [comment]
I'll add some server side doc updates to the PR then.

**Comment by felixbarny** [comment]
I can propose a section for the troubleshooting guide tomorrow.

**Comment by alvarolobato** [comment]
We've had another look at the risks involved in this change and went through these topics:

- Although we already have this situation with transactions, spans are much more diverse and users can use it in a very different way, making it easy to get into problems like the ones stated above by @simitt ie. a developer from one of the applications sets a key that changes on each span which could potentially break all the other applications
- We don't have a way to give feedback to the user other than the server logs. The result is that many documents would be discarded and the user could see a substantial drop of data in Kibana.
- We don't really know the full consequences of hitting the limit. Would all the span inserts fail or only the ones that try to add a new field? Are there any other consequences?
- There is no foreseeable solution right now with the amount of investigation we have done and we may need to change the model.
- We don't officially support OpenTracing yet and are focusing on getting Distributed Tracing and Metrics out, in which case deferring this wouldn't be a big problem.
- Not merging this only means that the method that adds the tags to the span would be NOOP, but won't block the agents from start working on supporting OpenTracing and have a fairly high support of the API until we find a better solution for this mapping.

For the above reasons we think we should wait until more investigation is made and avoid assuming this risk, potentially merging this in 6.5 but not in 6.4 as we initially intended.

**Comment by felixbarny** [comment]
I had a look at the storage model of the Zipkin Elasticsearch storage. It seems they have [disabled (`"enable": false`)](https://github.com/openzipkin/zipkin/blob/master/zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/VersionSpecificTemplates.java#L124) the tags, meaning they are not indexed, only stored.

@xeraa you said you where involved in that. Do you know if the reason to not index tags was to avoid mapping explosions?

I think having indexed tags would enable some powerful analytics use-cases but I can see that it comes with a risk.

**Comment by xeraa** [comment]
Shard explosion might also be a factor, but I think dots in tags are the main reason (since Zipkin is still supporting 2.x), see https://github.com/openzipkin/zipkin/pull/1685.

BTW we will probably need to dedot tags, otherwise `foo` and `foo.bar` will clash (is the type a string or an object). But `foo` and `foo_bar` will work. Metricbeat is already doing that (I think for collecting JSON and others).

**Comment by felixbarny** [comment]
Now that https://github.com/elastic/apm-server/pull/1156 has been merged, can we close this?
